### PR TITLE
Add backend tests from tarball using Tar.jl

### PR DESCRIPTION
**Pull request by ordicker**
Hi, 

ONNX have huge [test suite](https://github.com/onnx/onnx/blob/main/docs/OnnxBackendTest.md)
I wrote one test for review. 

I think we should aim to pass all the tests.

Same PR but with tarball (uncompressed)

### PR Checklist

- [X] Tests are added (only tests)
- [X] Documentation, if applicable

### CODE REVIEW

**Comment by Pangoraw** [comment]
Just my random opinion but it may be a good use case for an [Artifact](https://pkgdocs.julialang.org/v1/artifacts/) ? Putting the tar in the repo means that every user of ONNX.jl will need to download it even if not running tests.

**Comment by Pangoraw** [comment]
One problem that might arise is where to host the Tar archive. A solution which has been done before is to create a new repo where we create GitHub releases and upload the tar archives as artifacts there (see [NodeJSBuilder](https://github.com/davidanthoff/NodeJSBuilder/releases/tag/v18.16.0%2B0) for example). However, I don't know what is the most convenient for the FluxML org.

**Comment by ToucheSir** [comment]
If this tarball is being derived from the ONNX repo, would it make sense to use it directly via a submodule? Not sure how good Pkg support for those is.

**Comment by ToucheSir** [comment]
There's nothing fundamentally wrong with using artifacts, but they do require versioning and ongoing maintenance. Artifacts are also generally preferred for handling binary data or data that would be inappropriate to version control with git. My perhaps incorrect understanding of the ONNX tests is that they're not explicitly versioned, and everything is a VCS-compatible text file already. Thus it may make more sense to use a submodule for them instead of going through the longer process of creating and registering artifacts.

**Comment by ordicker** [comment]
Got it.
Could you help me setup the sub module?
I never worked with one.

**Comment by ordicker** [comment]
Do I need to make a new repo? 
and add it using `git submodule add <new-repo-url>`
@ToucheSir

**Comment by dfdx** [comment]
Sorry for the silence! I don't know much about git submodules either, but the following seems to work:

```
dfdx@dfdx:~/work/ONNX.jl$ git submodule add git@github.com:onnx/onnx.git test/onnx
Cloning into '/home/azbs/work/ONNX.jl/test/onnx'...
remote: Enumerating objects: 42908, done.
remote: Counting objects: 100% (834/834), done.
remote: Compressing objects: 100% (417/417), done.
remote: Total 42908 (delta 462), reused 692 (delta 403), pack-reused 42074
Receiving objects: 100% (42908/42908), 29.19 MiB | 309.00 KiB/s, done.
Resolving deltas: 100% (25262/25262), done.

dfdx@dfdx:~/work/ONNX.jl$ git status
On branch test-suites-as-submodule
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	new file:   .gitmodules
	new file:   test/onnx
```

**Comment by ToucheSir** [comment]
Yes, sorry. I saw your questions too, but I haven't yet looked for an answer on how to handle submodules in the usual build/CI workflow. Looking through https://github.com/JuliaLang/Pkg.jl/issues/708, perhaps a [subtree](https://www.atlassian.com/git/tutorials/git-subtree) would be a better option than a submodule?

**Comment by ordicker** [comment]
That's looks bad. 

I think we are overengineering the solution. It's 22MB of binary files (could be compressed, but that will add a pkg dependency).
I don't think we should trace the file itself, but the location should be public and owned by FluxML.

For any git base solution (submodule/subtree/ data repo), I need help opening that one under FluxML (just empty repo with meaningful name). 
Other solution, If FluxML has a server, then we could upload the file and use it as an artifact. 
I could upload it to my own repo (ordicker/onnx_backend_testing) but I think it should be under FluxML

**Comment by ToucheSir** [comment]
My main question with self hosting would be how to keep these files up to date and who is responsible for doing so. Also you said these were binary files? Are those created from the text files in the ONNX repo, or do they come from somewhere else?

**Comment by ordicker** [comment]
I don't we need to update them often (even at all). 
They are generated by [onnx](https://github.com/onnx/onnx) repo for testing. [link](https://github.com/onnx/onnx/blob/main/docs/OnnxBackendTest.md)
I generate the testing files (protobuf .pb) bundle it using tarball.

**Comment by ToucheSir** [comment]
Oh I see, they're already binary files. What I meant by updating is, what happens when the ONNX maintainers decide to change the contents of e.g. https://github.com/onnx/onnx/tree/main/onnx/backend/test/data/node/test_adagrad_multiple on their end? I presume we'd want have those updates on our end too, but because Julia artifacts are generally considered a snapshot of some data in time I think it'd require manual effort to do so.

**Comment by ordicker** [comment]
I’m proposing do it manually.
Mostly because it doesn’t change often.

**Comment by ordicker** [comment]
BTW, what about onnx.proto3. 
We don’t keep track on it either…

**Comment by ToucheSir** [comment]
I'm afraid I don't know what onnx.proto3 is. Does the official repro provide both protobuf v2 and v3 of the artifacts we're interested in, but we're only handling and/or able to handle the v2 versions right now?

**Comment by ordicker** [comment]
[onnx.proto3](https://github.com/FluxML/ONNX.jl/blob/master/src/onnx3_pb.jl) is the protobuf interface definition. 
We use it to generate (automaticly using ProtoBuf.jl) [onnx_pb.jl](https://github.com/FluxML/ONNX.jl/blob/master/src/onnx3_pb.jl)
I think onnx.proto3 support both v3 and v2, but I'm not sure.

I think we should focus on adding more operations, and the test suite is just scaffolding. 
When we implement most of the interface, we can remove it.

File: test/backend.jl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            f->filter(contains("input"),f).|>
+            pb_to_array
+        ## load the model
+        model = load(dirname*"/model.onnx",inputs...)
+        ## run it
+        return play!(model,inputs...)
+    end
+
+    @testset "Nodes" begin
+        prefix = extract("backend_data.tar")*"/data/node/"
```
Comment (dfdx) [inline test/backend.jl:60]: Are these files deleted after the test? If not, we should probably add the while directory to the gitignore.

**Comment by dfdx** [comment]
I don't like the overengineering discussion we are falling into, but adding a 22Mb file to the repo makes me nervous too. The problem is that once this files gets into the git history, there's no way to delete it from there. Even worse, when we ever change the file (and I assume _sometimes_ we will do it), the size is multiplied. This way we may end up with > 100Mb repo from which we only 136Kb (current size of `src`) is actually useful for end users. 

How about downloading the tar file directly from the [Github release](https://github.com/onnx/onnx/releases)? I believe we can stick to the [current release](https://github.com/onnx/onnx/archive/refs/tags/v1.14.0.tar.gz), download it in the beginning of the test suit and add the file to gitignore. I can foresee some edge cases in this approach too, but it looks like a little and totally reversible change that will unblock the real value - adding and testing more operations.

File: test/backend.jl (hunk 1 above)
Comment (ordicker) [reply test/backend.jl:60]: Good point, I though that if it is going to `/tmp/` I don’t have to worry about it, but you are right.
I will fix that.

**Comment by ordicker** [comment]
100% agreed with you.
Github release sound good to me.
Could you help me setup it up?

**Comment by dfdx** [comment]
I mean using the release of the onnx repo. That is, instead of putting the tar file to ONNX.jl, use something like this:

```julia
const ONNX_RELEASE_URL = "https://github.com/onnx/onnx/archive/refs/tags/v1.14.0.tar.gz"

onnx_release_tar_path = dirname(@__FILE__) * "/tests/backend_data.tar.gz"
onnx_release_path = dirname(@__FILE__) * "/tests/data/"
if !isfile(onnx_release_path)
    download(ONNX_RELEASE_URL, onnx_release_tar_path)
    extract(onnx_release_tar_path, onnx_release_path)
end
...
```

**Comment by dfdx** [comment]
Oh, I didn't realize it. Yet, I believe we can gnerate the files during the first run of the tests and keep the result on that directory.

**Comment by ordicker** [comment]
Can I assume that everyone has `python` ? 
And I think installing [protobuf](https://github.com/onnx/onnx/blob/f8da6749be56401880241221d01ffeb04c9cac95/requirements.txt) was a bit painful.

**Comment by dfdx** [comment]
Ouch. Now I see your your point. I agree then that putting a single tar into a repo is just fine, let's do it and come back to the question when we want to update it.

**Comment by dfdx** [comment]
Sounds good. Let's add the unpacked files to `.gitignore` and we are good to go.

**Comment by ordicker** [comment]
Bump @ToucheSir @dfdx

**Comment by dfdx** [comment]
Oh, sorry, I didn't realize you are awaiting for the confirmation :facepalm: Yes, the plan sounds great to me.

**Comment by ordicker** [comment]
So, are we done? @ToucheSir @dfdx 
I can’t merge it myself

**Comment by dfdx** [comment]
Damn! I missed it again! Sorry for that - I've been working in an extreme multitasking mode for the past few months, so if I don't reply quickly, don't hesitate to ping me again. 

The code looks good to me. I've also checked the artifact itself and verified that tests don't expose any obvious remote code execution capabilities. Thank you for all the patience with us!

**Comment by ordicker** [comment]
No need to apologize, day job is the worst 🤪

Now my plan is to add one operator at the time and hopefully the community will join
